doc/qos/Shared-headroom-pool-design.md

```diff
@@ -0,0 +1,235 @@
+# Shared headroom pool design
+
+## Table of Content
+
+### Revision
+
+### Scope
+
+This section describes the scope of this high-level design document in SONiC.
+
+### Definitions/Abbreviations
+
+This section covers the abbreviation if any, used in this high-level design document and its definitions.
+
+### Overview
+
+Currently, the dedicated memory are reserved as headroom for all PGs and can not shared by other traffic. Only all PGs suffer congestion at the same time can all the headroom buffer be occupied. However, this situation can hardly happen, which means all buffer is not used.
+
+The shared headroom pool is introduced to reduce the total memory occupied by headroom and thus enlarging the size of shared buffer pool.
+
+Currently, the shared buffer pool is calculated by subtracting reserved buffer of the following parts from the total available memory:
+
+- reserved buffer for each port on ingress and egress side.
+- reserved buffer for each PG, including xon and xoff. So it can be divided into two parts:
+  - reserved buffer for xon of all PGs
+  - reserved buffer for xoff of all PGs
+- reserved buffer for each queue.
+- reserved buffer for system, eg. management pool.
+
+In the shared headroom pool solution, the xoff buffer will no longer be reserved for each PG.
+
+In Mellanox's platform, a dedicated shared headroom pool will be introduced for xoff buffer. The size of this pool is reflected by `BUFFER_POOL|ingress_lossless_pool.xoff` and will be passed to SAI through `SAI_BUFFER_POOL_ATTR_XOFF_SIZE` attribute in `set_buffer_pool_attribute` API. The size can be determined in the following ways:
+
+- the over-subscribe ratio approach: the accumulative sizes of xoff for all PGs divided by over-subscribe ratio:
+- the congesting probability approach:
+  1. specify the probability of congestion for each lossless buffer profile
+  2. put the product of all PGs together
+- the congesting factor approach: the congesting factor represents how many ports can suffer congestion at the same time at most.
+
+All the ways are configurable. The shared headroom pool won't be enabled if none of the configuration provided.
```

Comment:
can you specify explicitly that the 'oversubscribe ratio' approach is chosen? Also we allow to configure the size as well.

Reply:
Yes, I will.
BTW I am also working on merging this doc into the original doc for dynamic buffer calculation as this is based on it.
I believe it will be more clear after that